restler/utils/logging/trace_db.py

```diff
@@ -0,0 +1,225 @@
+
+    if threadLocal.sequence_trace.combination_id != combination_id:
+        raise Exception("ERROR: Combination ID must match Combination ID of initialized trace log.")
+
+    threadLocal.sequence_trace.request_id = request_id
+    threadLocal.sequence_trace.tags.update(tags)
+
+def get_trace_log():
+    return getattr(threadLocal, 'sequence_trace', None)
+
+def clear_trace_log(request_id=None, combination_id=None):
+    if combination_id is None:
+        raise Exception("ERROR: Sequence ID must be specified.")
+    if threadLocal.sequence_trace.combination_id != combination_id:
+        raise Exception("ERROR: Sequence ID must match sequence ID of initialized trace log.")
+
+    if request_id is not None:
+        threadLocal.sequence_trace.request_id = None
+    elif combination_id is not None:
+        threadLocal.sequence_trace = None
+
+def log_request_response(request=None, response=None,  tags={}, timestamp=None):
+    _db.log_request_response(request=request, response=response, tags=tags, timestamp=timestamp)
+
+def get_sequence_id():
+    return _db.get_sequence_id()
+
+
+class TraceDatabase:
+    """ This class enables structured storage and retrieval of RESTler logs.
+        The serialization format is pluggable and may be specified by the user (currently,
+        only one logger at a time is supported).
+        The default format is newline-delimited json.
+    """
+    def __init__(self, storage_writer):
+        self.storage_writer = storage_writer
+        self._log_queue = queue.SimpleQueue()
+        self._finished = False
+
+    def log(self, message):
+        if self._finished:
+            raise Exception("ERROR: log() should not be called after finish().")
+        self._log_queue.put(message)
+
+    def log_queue_empty(self):
+        return self._log_queue.empty()
+
+    def finish(self):
+        if self._finished:
+            raise Exception("ERROR: finish() should only be called once.")
+        self._finished = True
+        self._log_queue.put(None)
+
+    def save_log_message(self):
+        """Takes a message off the queue and saves it using the specified serializer.
+            If fewer messages are available, it will save all available messages.
+        """
+        message = self._log_queue.get()
+        if message is None:
+            # End of tracing signal
+            return
+        self.storage_writer.save(message)
+
+    def get_sequence_id(self):
+        tls_trace_log = get_trace_log()
+        if tls_trace_log is None:
+            return None
+        return tls_trace_log.sequence_id
+
+    def log_request_response(self, request=None, response=None,  tags={}, timestamp=None):
+        if request is None and response is None:
+            raise Exception("ERROR: Request or response must be specified.")
+        try:
+            tls_trace_log = get_trace_log()
+            if tls_trace_log is None:
```

Comment:
For my own understanding, how does this condition work (checking if `get_trace_log` returns None to determine if the request is outside the sequence context)? Do checkers run in their own thread (that do not initialize `sequence_trace`)?

Reply:
Garbage collection runs in its own thread.  Checkers do not run in their own thread - they run after the "main" sequence has finished.  However, checkers in general can have custom logic and may decide not to run 'initialize_sequence_trace', which this logic would cover.

Reply:
This PR iteration does not yet have the checker logic fully working for tracking sequences.

Reply:
The checkers are invoked after the main driver sequence has finished running (see driver.py line 210-217).  So first the main driver will initialize and clear the sequence trace (see 'clear_trace_log'), and then the checker will start a new sequence trace (since the checker will re-render each sequence to which it is being applied from scratch).  

The main driver should always clear the trace log before passing control to the checker.  I will try to rename 'clear_trace_log' to better reflect the above intent.

Reply:
Ahh okay, thank!